Update Mixer documentation

kevincwells opened the pull request:
Updating the documentation for mixer to account for recent changes,
and to better clarify some concepts. This is to coincide with an
upcoming release of mixer, in which several important changes will
be introduced.

1) The mixer CLI has been rewritten, supporting POSIX-style flags
and a modern command layout. This is a breaking change, as commands
and flags have been renamed.

2) The initialization steps for mixer have been further automated,
so the documentation has been updated to account for the new,
simpler approach.

3) Documentation has been added for the recent command for
automating the addition of bundles to a mix.

4) The tool now handles caching upstream CLR bundle definitions
in a more automated way, and so the 'mixer get-bundles' command
(and its corresponding documentation section) has been removed.

Signed-off-by: Kevin C. Wells <kevin.c.wells@intel.com>

kevincwells commented:
@tmarcu -- please read this over and verify its accuracy.  The diff might not be incredibly useful (effectively the entire file has been changed at least slightly), so you might want to just read the final file.

This PR has a lot of little clean-up throughout, trying to consistently use the :file:\`path/to/file.ext\` and other inline formatting techniques. That said, I notice clearlinux.org doesn't actually take advantage of many of these, as the stylesheets render much of these modified texts the same.

This builds and appears to render correctly for me, but I am not pulling in all the final stylesheets used on clearlinux.org.

rcaballeromx commented:
@kevincwells Thank your for this update. I'll review and provide comments. Regarding the inline markup: Not all classes have been implemented on the website's CSS but they are imperative for our automated document based testing efforts. Thank you for using them.

tmarcu commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -70,195 +83,220 @@ Mixing
+
+   Behind the scenes, mixer uses a local cache of the upstream Clear Linux
+   bundle definitions. These are stored in the 
+   :file:`.mixer/upstream-bundles/clr-bundles-{VER}/bundles/` directory in your
```
@kevincwells I'm sure this doesn't need to be said, but this kind of content will have to be updated as the new bundle/folder PR changes go in soon :)

matthewrsj commented:
bump

tmarcu commented:
@rcaballeromx @ECouzens What is going on with this PR...?

ECouzens commented:
@tmarcu apologies . We're waiting for an additional SW release this week before we merge this PR. @bktan8 will verify when that's released and will comment here when we are go for merge

tmarcu commented:
:grinning:

kevincwells commented:
@rcaballeromx @ECouzens  -- There have been new additions to the `mixer` tool since this PR was added. I'd like to update the docs to reflect them, but I'm a bit stuck with this PR still outstanding.

Would you prefer I:
1. Update this branch (force push) with new content
1. Add a new commit to this branch with new content
1. Start a new branch (and submit a new PR) that is, temporarily, based on top of the commits in this branch, and then rebase on top of `master` once this is merged?

I don't want to mess up your validation/merge process, so I won't make any changes to this branch without your approval.

mrkz commented:
@rcaballeromx ^

rcaballeromx commented:
@kevincwells The correct answer is option 2. Just as you would in the code, please make any updates on this content on this branch but in different commits. Thanks!

tmarcu commented:
The new Mixer is now released in upstream, so this documentation has to get in ASAP. @kevincwells is this already ready to push on your end?

kevincwells commented:
@rcaballeromx -- I'm definitely willing to do that, but wouldn't that effectively invalidate the two approvals and other review that has gone through on this PR? The new commits would need to go through a new approval. It's a functional solution, but 

Is there actually still something preventing this commit from being merged?

kevincwells commented:
@tmarcu -- It's about half done and is my top priority today.

rcaballeromx commented:
@kevincwells You are absolutely correct, just as it would with code. I haven't merged this content because based on the comments, it is clear that it is not complete.
As soon as you complete the content, I will review. This is also my top priority for today.

bktan8 commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -70,211 +92,267 @@ Mixing
       VERSIONURL=<URL where the version of the mix will be hosted>
       FORMAT=1
 
```
This builder.conf is missing the [Server] and [Mixer] sections:

[Mixer]
LOCAL_BUNDLE_DIR=/home/clr/mix/local-bundles

[Server]
debuginfo_banned=true
debuginfo_lib=/usr/lib/debug/
debuginfo_src=/usr/src/debug/

kevincwells replied:
@bktan8 Thanks for taking a look at this.  Yes, this is one of the changes coming in the update for 4.0

bktan8 commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -70,211 +92,267 @@ Mixing
+
+   This command copies the specified bundle defintion files from your
+   configured upstream version of Clear Linux (:file:`.clearversion`) into
+   your :file:`mix-bundles` directory.
```
I don't see the "mix-bundles" directory in my testing.  I see a "mixbundles" file which contains a list of bundles.

kevincwells replied:
Again, yes, while these docs have been valid for the last month, _new changes_ have been made in that time that make much of this outdated. I'm updating the documentation as we speak.

tmarcu commented:
@bktan8 @rcaballeromx please do not review this PR any more until the new commits are in.

This was complete with multiple approvals, but it got mangled with absolutely irrelevant commits and then left to get stale for a month.

bktan8 commented:
got it!

rcaballeromx commented:
@tmarcu Understood. I won't review until you or @kevincwells let me know.

kevincwells commented:
@rcaballeromx @ECouzens @bktan8 -- I've pushed a new commit with the updates to cover what's new in Mixer 4.0.

@ECouzens -- I liked what you were trying to do with the jump-to-label tag for "Step 4", but I had to remove it, because it broke the auto-numbering for the list. It made that section start a new count, so that section was no longer being numbered "4.", but instead "1."

kevincwells commented:
@tmarcu - nice catch. Just force pushed a fix

mltullis commented:
Status: @bktan8 is reviewing/validating and will hand off to @ECouzens for a final look and merge.

kevincwells commented:
@bktan8 and others: I've had to (force) push a minor change to this commit. The names of the local rpm and yum repos were wrong, and the definition of "basic" bundle validation was no longer correct.

bktan8 commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,278 +3,528 @@
+*************
 
-To start working with the Mixer tools, you'll need a recent image of Clear
+To start working with the mixer tools, you'll need a recent image of Clear
```
@tmarcu @kevincwells: Are we purposely choosing to lowercase the "m" in Mixer?

kevincwells replied:
@bktan8  This is actually a change @ECouzens  made in Commit 5b4690f9d32a8be07149bd549d9aa7a744829c95

bktan8 replied:
I am wondering if we should use "Mixer" since it's a product.

rcaballeromx replied:
@bktan8 It is incorrect grammar to capitalize non-proper names. If the tool's name was trademarked, we would have to follow the trademark capitalization but that is not the case with this feature.

tmarcu replied:
I defer to @rcaballeromx for grammar rules :smile:

bktan8 commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,278 +3,528 @@
-      #. Make any needed modifications to the bundle set.
-      #. Commit the result::
+   Local bundle definition files live in the :file:`local-bundles` directory.
+   The locaiton of this directory is specified by ``LOCAL_BUNDLE_DIR`` in your
```
@kevincwells : "locaiton" is mispelled.

kevincwells replied:
I've fixed this in a forthcoming push. Just waiting to see if there are other typos to fix :)

rcaballeromx replied:
@kevincwells Please push your latest updates and I'll take one final look and fix minor issues for you before merging. If you push your changes in the next couple of hours, I can merge before EoD.

kevincwells replied:
@rcaballeromx Great! I just pushed the change that fixed this typo. It is the only change I am currently aware of. Feel free to fix any other minor things like that as you see them :) Thanks!

bktan8 commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,278 +3,528 @@
+   .. code-block:: bash
 
-   To define your bundles:
+      # mixer bundle list upstream
```
@kevincwells - is there a way to get a tree view of one bundle (upstream or local)?

kevincwells replied:
Of a single bundle? No, not with the current commands.

The ability to "inspect" a single bundle is in the backlog, but is lower priority than other things we're working on right now. For that hypothetical `mixer bundle inspect` command, I wanted to show more than just a bundle hierarchy, but actually help the user visualize what packages are pulled in, too.

bktan8 commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,278 +3,528 @@
+   this is currently an experimental feature, you can (and should) use the new
+   chroot-builder by passing the :option:`--new-chroots` flag. The legacy 
+   chroot-builder will soon be deprecated, and mixer will use the new version
+   automatically.
```
@kevincwells, regarding:
+   Mixer now has a new chroot-builder built into the mixer tool itself. While
+   this is currently an experimental feature, you can (and should) use the new
+   chroot-builder by passing the :option:`--new-chroots` flag. The legacy 
+   chroot-builder will soon be deprecated, and mixer will use the new version
+   automatically.

So if you don't specify the --new-chroots flag, which chroot builder is using by default?  The new experimental one or the legacy?  And if it's the new one by default, why "should" you specify the flag?

kevincwells replied:
>So if you don't specify the --new-chroots flag, which chroot builder is using by default? The new experimental one or the legacy?

A couple lines earlier (L381), the paragraph starts with "By default, mixer will use the legacy chroot-builder"

bktan8 commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,278 +3,528 @@
 
-   Edit this to include  all the bundles you want pre-installed into your
-   image. For a minimal, base image this would be::
+      # cp /usr/share/defaults/ister/ister.json relase-image-config.json
```
@kevincwells, I think "/usr/share/defaults/ister/ister.json" is not a good example for users for a number of reasons:
- "disk:" is set to "/dev/sdb"
- "Bundles:" doesn't show the minimal bundles
- "DestinationType" is set to "Physical"  

I would prefer to keep using https://raw.githubusercontent.com/bryteise/ister/master/release-image-config.json or modify "/usr/share/defaults/ister/ister.json" to match.  

I'm just thinking it's good to make the example simple and the way we intend it to be, especially for new users.

kevincwells replied:
I don't think we can say what way "we intend it to be" -- the whole point of `ister`'s config is that it can be used to build for any number of different formats, and those will be entirely dependent on the user. The `release-image-config.json` you mention (which we still link to in the sentence immediately following the lines mentioned here) is not necessarily representative of the image format any given customer will want. The point of this command is to give them a starting point for the syntax.

I'm open to changing the default config that is shipped with `ister` (though that's a different project and would need to be made there), but we deliberately moved away from telling the user to curl a file from github for this step.

bktan8 commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,278 +3,528 @@
+Creating your next mix version
+==============================
+
+**Update the next Mix version info**. Update the :file:`.mixversion` file to
```
@kevincwells - .mixversion should be just mixversion.

kevincwells replied:
Oh, I knew I was forgetting one of the many new commands. This section needs to be updated to instead use the new `mixer versions update` command.

kevincwells commented:
@rcaballeromx -- @bktan8 has pointed out a couple of places that need updated. I can submit a new push in the next half hour that fixes these things. I'll post here when I do, so please be sure to pull the latest version.

bktan8 commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,278 +3,528 @@
+   RPM files and not corrupt.
+
+#. **Add/remove/edit/list the bundles in your mix**. The bundles in your mix are
+   specified in the Mix Bundle List. This list is stored as a flat file called
```
@kevincwells - is there a reason M, B, and L are capitalized in "Mix Bundle List"?

kevincwells replied:
I don't want to get into a grammar argument about what constitutes a "proper noun", but the short version is that I was treating it as one, a) to improve clarity and help people understand that this is a specific, unique thing, and b) to be consistent with the way it is written throughout all of the new command help doc information.

rcaballeromx replied:
I can do both a without the capitalization. Sadly, b means that the help doc must be revised at a later date. As any writer can attest, you cannot fight grammar. I'll make sure to mark it clearly to achieve the desired clarity and specificity.

kevincwells replied:
Is there actually a reason to remove these capitalizations, here and throughout the commands and code comments? Just to be completely clear, the capitalization is _not_ grammatically incorrect. It _might_ be against our adopted style guide (do we actually have one?), but that's a different concept altogether.

rcaballeromx replied:
@kevincwells Yes there is. The dictionary is clear: https://en.oxforddictionaries.com/spelling/using-capital-letters
We have had a style guide for months. You can find the capitalization rules here: https://clearlinux.org/documentation/clear-linux/reference/collaboration/documentation/grammar
I will follow both the Oxford grammar and our style guide in my final tweaks.

kevincwells replied:
@rcaballeromx 

Thanks for linking me to the style guide -- when I asked if we have one, that was indeed a serious question. I have bookmarked the page and will do my best to adhere to it.

In regards to the style guide and the OED link -- I see nothing at all at either link that disputes this capitalization. The OED link is specifically a list of situations in which "you should always use a capital letter" -- it makes no assertions about when you should not. To my surprise, it makes no mention of the word "proper noun", which is often one of the cases cited. The style guide, however, does state that. "Mix Bundle List" is, in fact, a proper noun.

OED's definition for [proper noun](https://en.oxforddictionaries.com/definition/proper_noun) is quite open, defining it through contrast to a [common nun](https://en.oxforddictionaries.com/definition/common_noun): "A noun denoting a class of objects or a concept as opposed to a particular individual."

[Merriam-Webster](https://www.merriam-webster.com/dictionary/proper%20noun), [Dictionary.com](http://www.dictionary.com/browse/proper-noun), and [Wikipedia](https://en.wikipedia.org/wiki/Proper_noun) make the distinction in both their definition of proper noun and common noun: a "proper noun" is a noun that refers to a unique entity, while a "common noun" refers to a non-unique instance from within a class of entities.

While there are many lists, there is only one Mix Bundle List.

This same definition says that we should, in fact, actually capitalize "Mixer" when we are referring to it as the entity: there are many mixers (many things that mix other things), but there is one Mixer (the tool to build Clear Linux update content). And that would 100% be in keeping with the style guide. Where we _should not_ capitalize "mixer" (and the style guide addresses this when referring to not changing the case of software components) is when it is a literal command, referring to the lowercase-m binary executable.

As such, it would be appropriate to write: "To begin use Mixer, run `mixer init`."

rcaballeromx commented:
@kevincwells @bktan8 My day is almost over and I do need to change some markup and minor punctuation. Make your changes and I will pull them first thing Monday morning to make mine and have them ready by the time you get in the office.

kevincwells commented:
@rcaballeromx @bktan8  I've added the clarification to `mixer bundle remove`, and added the new section for the `mixer versions update` command. I've also run additional spellchecking and caught a few more.

Let me know if you run into any issues with your final tweaks tomorrow.

kevincwells commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,278 +3,577 @@
+
+   If you are creating RPMs from scratch, you can use ``autospec``, ``mock``,
+   ``rpmbuild``, etc. to build them. If they are not built on Clear,
+   make sure your configuration and toolchain builds them correctly for Clear,
```
That is a good question.  I didn't change the wording here (it appears as new lines, but it's just been reformatted) because I don't know enough about Clear's RPM requirements to articulate it.

Could @matthewrsj @phmccarty or someone elaborate on what could go here?

tmarcu replied:
We can link to https://github.com/clearlinux/common for this section. I don't want to explain any of it in here though, because it will make it too long and off topic.

rcaballeromx replied:
@tmarcu I agree, the link is sufficient.

rcaballeromx commented:
@kevincwells I am frankly not that worried about the capitalization issues. I have been fixing them as I go while editing and you can take a look at the final draft. Content-wise however, the text has more serious issues like excessive use of passive voice and relative pronouns, incorrect cross-referencing, and inconsistent use of terminology. I am working on a draft fixing all these issues but it is taking a lot longer than expected due to the complexity of the content. I am working hard at it to have a draft uploaded by today EoD.

kevincwells commented:
@rcaballeromx :grinning:  I wholeheartedly believe you and appreciate you fixing those things.

rcaballeromx commented:
@tpleavitt Thank you for the comments. @ECouzens is performing an additional copy edit to clean up all those typos and artifacts that get introduced by massive rewrites. @kevincwells @tmarcu Can you take a look and ensure the content is still technically accurate? I tried to keep the technical information intact while providing more explicit command examples. Lastly, keep in mind that all '#' command prompts were replaced with sudo to prepare the document for automated testing.

kevincwells commented:
@rcaballeromx I'm reviewing now. Very happy with the structural changes -- great work. I'm doing the technical comments as part of a formal "Review", so those will come in a few minutes when I'm done. :smile:

kevincwells commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+
+If you have been tracking your workspace with Git, you can restore the mix to
+an earlier state but be careful of "rewriting history" if you are publishing
+the mix content to users already.
```
You removed the comma before "but". Without it, this is a run-on sentence.

Requested edit: "state but" -> "state, but"

rcaballeromx replied:
Done by Evan

kevincwells commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+
+.. code-block:: bash
+
+   sudo mixer versions update --upstream-version 21070
```
Requested edit: remove `sudo`

rcaballeromx replied:
done

kevincwells commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+
+.. code-block:: bash
+
+   sudo mixer versions update --upstream-version latest
```
Requested edit: remove `sudo`

rcaballeromx replied:
done

kevincwells commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+
+This command sets the upstream version to the latest released version of
+upstream |CL| within the same format version. The :command:`mixer
+versions update` command does not allow to set an upstream version to a
```
You removed the word "you", but without it, "The `mixer versions update` command does not allow to set" is incorrect.

rcaballeromx replied:
Added 'you' back

kevincwells commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+"format bump" build, which is currently a manual process. See :ref:`mixer-format`
+for more information.
+
+Learn which mix version or upstream version you currently are on with the
```
Similar to my comment about `mixer bundle validate`, the active voice (and thus command form) here kind of loses that this is an optional step. Not strictly wrong, it just feels weird.

rcaballeromx replied:
Added 'Optionally, you can" to communicate the fact this is optional.

kevincwells commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+
+.. code-block:: bash
+
+   sudo mixer versions
```
Requested edit: remove `sudo`

rcaballeromx replied:
Done

kevincwells commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
 
-   Edit the config file to include all bundles that you want *preinstalled*
-   into your image. The rest of the bundles in your mix will be available to your users via:
+   sudo mixer bundle remove bundle1
```
Requested edit: remove `sudo`

rcaballeromx replied:
done

kevincwells commented:
@rcaballeromx Ah, I see @ECouzens just pushed a few new commits on top of yours that I was reviewing. This makes Github mark the _vast majority_ of my edits as "outdated" (and hidden by default). Please do not miss these edits, as they are still valid.

kevincwells commented:
@ECouzens -- I notice this branch now has a few merge commits (e.g. 888cf2ff27733b0481de2367ee1fbffa66a2048a), in addition to the edit commits. I'm guessing you did a `git merge` into your own local copy of this branch, then pushed these commits along with your edits on accident.  As is, this branch needs to be rebased to remove them. In the future, can you be sure to either pull the branch from scratch, do a `git pull` instead of a `git merge`, or manually rebase your changes before pushing them?

kevincwells commented:
@rcaballeromx and others:

I'd like to reiterate that the review process I've observed here is not how Git (or GitHub) are supposed to be used. When I submit a pull request, the commits live on my personal branch in my personal GitHub fork repository (in this case, `kevincwells/mixer-tools`). In general, no one should write to this branch except me. This is not a communal branch on a communal repository like the original, it is a personal fork of that original.

If there are changes you would like before my PR can be merged, and I can reasonably make them myself, you should note the changes as part of GitHub's "Request changes" review feature. It would then be my responsibility to make those changes on my personal branch, and then push (or force-push) to update my PR.

The fact that I can (and likely will need to) force-push changes to commits should illustrate why others should not be pushing to my branch. If I am making an edit to my commit, and some other person pushes additional commits to it before I finish, I can accidentally erase them by force-pushing my update. (Yes, this is a risk any time people are sharing a branch and editing it in parallel. But again, this _is not_, in fact, a shared branch.)

If your team would like to instead be responsible for editing the contents of my PR, perhaps the best process would be to have developers instead submit their PR's to a "review" branch on the upstream repository. Your team could then accept the PR and merge the commits into the review branch, where you would be free to make additional commits or edit those that already exist.

rcaballeromx commented:
@kevincwells Thank you very much for the feedback regarding this review process. In the past, we followed the process just as you described it and as Github was intended to work. However, that process creates an overhead. For example, if I review the content and comment on the changes providing the edits, you then need to implement said edits. Thus, both of us have performed the same work. I will take your feedback to our documentation team meeting and discuss alternatives and ways to avoid the issues.

rcaballeromx commented:
@kevincwells I will not miss your edits and thank you for making all the comments in such a precise way.

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
-   :file:`upstreamversion` file), and cleans up old versions once they are no
-   longer needed. You can see what upstream bundles are available by running:
+This command prints a tree view of the mix bundle list explicitly showing each
+included bundle.
```
I replaced the text with your last sentence. I agree your version provides better detail as to what the flag does.

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
 
-   With either of the above commands, you can pass the :option:`--tree` flag to
-   see a tree view of what other bundles each bundle includes.
+   sudo mixer bundle list upstream
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+.. code-block:: bash
 
-      # mixer bundle add bundle1 [bundle2...]
+   sudo mixer bundle list local
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
-   this information when you run :command:`mixer bundle list`.
+Both the local or upstream :command:`bundle list` commands accept the
+:option:`--tree` flag to print a tree view explicitly showing each included
+bundle.
```
Reused the wording from that comment here too.

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
-   throughout the system: it will check that the bundle syntax is valid and can
-   be parsed, and that the bundle file has a valid name. If you would like to
-   manually run this validation on a bundle, you can run:
+   sudo mixer bundle edit bundle1
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+   sudo mixer bundle edit bundle1
 
-   .. code-block:: bash
+If `bundle1` is found in your local bundles, mixer edits the bundle definition
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+In both cases, mixer launches your default editor to edit the file. When the
+editor closes, mixer automatically validates the edited bundle file and
+reports any errors found. If mixer finds an error, you can edit the file as-
+is, revert and edit, or skip and move on to the next bundle. If you skip a
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
-         editing a bundle definition file. All of the above :command:`mixer 
-         bundle` commands support an optional :option:`--git` flag that will
-         automatically apply a git commit when they are finished.
+   sudo mixer bundle edit bundle1 bundle2 [bundle3 ...]
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
 
-   .. code-block:: bash
+To create a totally **new bundle**, the bundle name you specify cannot exist
+upstream. If that is the case, create a `new-bundle` with the following
```
Not really obvious but already covered in the previous section regarding editing. I found it redundant to explicitly stating it here.

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+.. note::
 
-#. **Create update**. In the workspace, run:
+   The :command:`mixer bundle edit` accepts multiple bundles at once. Thus,
```
Done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
    .. code-block:: bash
 
-    # sudo mixer build update
+      sudo mixer bundle edit new-bundle1 new-bundle2 [new-bundle3 ...]
```
Done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
-   new swupd-server by passing the :option:`--new-swupd` flag. The legacy
-   swupd-server will soon be deprecated, and mixer will use the new version
-   automatically.
+Add `bundle1` to your mix easily with the following command:
```
Done.

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+.. code-block:: bash
 
-   .. code-block:: bash
+   sudo mixer bundle add bundle1
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+.. code-block:: bash
 
-   .. code-block:: bash
+   sudo mixer bundle add bundle1 bundle2 [bundle3 ...]
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+   sudo mixer bundle remove bundle1
 
-   .. code-block:: bash 
+This command remove the `bundle1` from the mix bundle list stored in the
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
 
-   Keeping this list small allows for a smaller image size. For a minimal,
-   base image, this list would be:
+   sudo mixer bundle remove --local bundle1
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
-   should be built from. ``ister`` allows you to build an image from any mix
-   version that you have built, not just the current one. For the first build
-   example we've been using, ``Version`` would be set to 10.
+   sudo mixer bundle remove --local --mix=false bundle1
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
 
-   The output from this should be an image that is bootable as a virtual
-   machine and can be installed on bare metal.
+Mixer performs basic **validation** on all bundles when used
```
Agreed. Done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
-      current format version by running:
+Mixer checks the validity of the bundle's syntax and name. Mixer also ensures
+the bundle can be parsed. Run this validation manually on bundle1 with the
+following command:
```
I removed the requested sentence and added the adverb "Optionally" to communicate the command is not required. I also added a note after the command with the example you provided of when the command might come in handy.

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+.. code-block:: bash
 
-         # cat /usr/share/defaults/swupd/format
+   sudo mixer bundle validate bundle1
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+.. code-block:: bash
 
-   .. code-block:: bash
+   sudo mixer bundle validate --strict bundle1
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+.. code-block:: bash
 
-   .. code-block:: bash
+   sudo mixer bundle validate bundle1 bundle2 [bundle3 ...]
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+commit after modifying the mix bundle list or editing a bundle definition file.
 
-   .. code-block:: bash
+All the :command:`mixer bundle` commands on the previous sections support an
```
I changed the preposition. The suggested edit with 'of' is controversial. ;) 
For consistency and simplicity I kept it out but if you want to learn more I have this very nice article about it:
https://jakubmarian.com/is-it-all-the-or-all-of-the-in-english/

kevincwells replied:
:wink: Yeah, that's why I made it was a suggestion. I'm happy with your changes (and lack of changes) here.

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+
+.. code-block:: bash
+
+   sudo mixer bundle remove --git bundle1
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+files.** Mixer automatically clears out any contents in this directory before
+populating it on-the-fly as mixer builds the chroots.
+
+We have build a new chroot-builder into the mixer tool itself. While this is
```
I changed the verb but incorporated is difficult to understand for non-native speakers. I replaced it instead with 'added' since it also avoids the confusion with build but is a simple word.

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+populating it on-the-fly as mixer builds the chroots.
+
+We have build a new chroot-builder into the mixer tool itself. While this is
+currently an experimental feature, you should use the new chroot-builder. To use
```
I leave that decision to you all but, for clarity, the documentation should not have ambiguous wording.

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+
+.. code-block:: bash
+
+   sudo swupd bundle add
```
Ok, so I was not seeing hypens. ;) That's reassuring.

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+   sudo mixer build image --format 1
+
+This command outputs an image bootable as a virtual machine and which can be
+installed on bare metal.
```
Removed.

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+
+.. code-block:: bash
+
+   sudo mixer versions update
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+
+.. code-block:: bash
+
+   sudo mixer versions update --increment 100
```
done

rcaballeromx commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+
+.. code-block:: bash
+
+   sudo mixer versions update --mix-version 200
```
done

kevincwells commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+  embedded in images created for your mix. The `swupd-client` looks at these
+  URLs to figure out if there is a new version available and the location from
+  which to download the updated content. Think of these links as the
+  equivalent to the |CL| `update page`_ but for your mix.
```
@rcaballeromx - Awesome -- that's great that we can actually point them to something we've written on the subject.

+1 on all your edits to this section.

I noticed (in my own version, and persisted in your edit) that I left out the word "the" in the line "If your web server is on **[the]** same machine as this directory".

kevincwells commented on source/clear-linux/guides/maintenance/mixer.rst:
```diff
@@ -3,577 +3,765 @@
+
+* The `FORMAT` variable relates to the format bump. To learn more about the
+  `FORMAT` option, see :ref:`mixer-format` and the `format bumps wiki`_. For
+  now, leave the `FORMAT` value unchanged.
```
Agreed that it's worth having this section in general. I'm satisfied with your reword solution.  +1

kevincwells commented:
+1 on all of @rcaballeromx's changes -- great job!

I found a missing word (my doing) and a very minor formatting issue, and just pushed a commit that fixes both.

The only outstanding things I am aware of are the two things I'd like @tmarcu and @matthewrsj to weigh in on:
- Am I correct that mixer only auto-generates the `Swupd_Root.pem` file if the location defined in `CERT` doesn't exist? @rcaballeromx made that change, but I'm not an expert on that part of the code.
- Should we, in fact, be instructing people to use the `--new-swupd` and `--new-chroots` yet?

tmarcu commented:
@kevincwells 
1.) Yes it will only generate if the file doesn't exist.
2.) yes we should be telling people to use the new flags.
